examples/chat.rs

```diff
@@ -0,0 +1,96 @@
+    let future = socket.incoming().for_each(move |(stream, addr)| {
+        let connections = connections.clone();
+        let handle_inner = handle.clone();
+        // We create a new future in which we create all other futures.
+        // This makes `stream` be bound on the `lazy` future's task, allowing
+        // `ReadHalf` and `WriteHalf` to be shared between inner futures.
+        handle.spawn_fn(move || {
+            println!("New Connection: {}", addr);
+            let (reader, writer) = stream.split();
+            // channel to send messages to this connection from other futures
+            let (tx, rx) = tokio_core::channel::channel(&handle_inner).unwrap();
+            // add sender to hashmap of all current connections
+            connections.borrow_mut().insert(addr, tx);
+
+            let connections_inner = connections.clone();
+            // https://users.rust-lang.org/t/loop-futures-for-client-handling/6950/2
+            // We have an endless loop reading from a client.
+            // In order to fuse the reading and writing futures in the end, we need to have the same
+            // output type. Therefore we use `(Option<BufReader<ReadHalf<TcpStream>>>,
+            // Option<WriteHalf<TcpStream>>)`.
+            let reader = BufReader::new(reader);
+            let socket_reader = stream::iter::<_, _, std::io::Error>(iter::repeat(()).map(Ok)).fold((Some(reader),None), move |(reader, _), _| {
+                let reader = reader.unwrap();
+                let connections = connections_inner.clone();
+                // read and parse length prefix
+                io::read_until(reader, '\n' as u8, vec![])
+                    .and_then(|(reader, vec)| futures::lazy(|| {
+                        // EOF was hit without reading a delimiter
+                        if vec.len() == 0 {
+                            futures::failed((std::io::Error::new(std::io::ErrorKind::BrokenPipe, "Broken Pipe"))).boxed()
+                        } else {
+                            futures::finished((reader, vec)).boxed()
+                        }
+                    }))
+                    // convert bytes into string
+                    .map(|(reader, vec)| (reader, String::from_utf8(vec).unwrap()))
+                    .and_then(move |(reader, message)| {
+                        println!("{}: {:?}", addr, message);
+                        // For each open connection except the sender, send the string via the channel
+                        for tx in connections.borrow_mut().iter().filter(|&(&k,_)| k != addr).map(|(_,v)| v) {
+                            tx.send(message.clone()).unwrap();
+                        }
+                        futures::finished((Some(reader),None))
+                    })
+            });
+
+            // Whenever we receive a string on the Receiver, we write it to `WriteHalf<TcpStream>`.
+            let socket_writer = rx.fold((None, Some(writer)), move |(_, writer), msg| {
```

Comment:
This is pretty gnarly, perhaps it could be written as:

```rust
let writer = rx.fold(writer, |wr, msg| {
    io::write_all(wr, msg)
});
```

and then transforms can happen later if need be.